rasa/shared/core/domain.py

```diff
@@ -448,6 +462,24 @@ def __init__(
         self.store_entities_as_slots = store_entities_as_slots
         self._check_domain_sanity()
 
+    @staticmethod
+    def _collect_overridden_default_intents(
+        intents: Union[Set[Text], List[Text], List[Dict[Text, Any]]]
+    ) -> List[Text]:
+        """Collects the default intents overridden by the user.
+
+        Args:
+            intents: User-provided intents.
+
+        Returns:
+            User-defined intents that are default intents.
+        """
+        intent_names: Set[Text] = {
```

Comment:
optional: What I don't like about it is that it replicates some behavior of `collect_intent_properties`:
- checking if we have a `dict` or `str`
- looping over the intents again (that's ok imo)

Another thing which isn't ideal imo:
- we have a lot of `staticmethod` methods which all manipulate the state of the domain. That's slightly paradox imo. 

```python
def __init__(....):
  ...
  # don't add the default intents here
  self.intent_properties = self.collect_intent_properties(intents, entities)
  # add them here and remember which ones were overriden
  self.intent_properties, self.overriden_default_intents = self.add_missing_default_intents(self.intent_properties)
```

```

Reply:
> checking if we have a dict or str

I thought about this but didn't see a reasonable way to abstract the `dict` or `str` checking to be re-used in those methods. it would have to be at the expense of introducing a new method `is_intent_of_dict_type` or something similar which is not necessary. it only happens twice so I think that's OK

> we have a lot of staticmethod methods which all manipulate the state of the domain. That's slightly paradox imo.

I would say that's appropriate in these cases as they're used in the constructor

> add them here and remember which ones were overriden

That makes the `self.overriden_default_intents` more complicated than it needs to be (it can just be a list of intent names, whereas `self.intent_properties` is an object). we'd have to return an awkward tuple